### ISSUE TITLE
Cannot create task

### ISSUE THREAD

**Issue by Nyholm** [symfony-tools/carsonbot#179]
This error shows up in the log: 

>  app.ERROR: Failed dispatching "pull_request" event for "carsonbot-playground/symfony" repository. {"exception":"[object] (Doctrine\\DBAL\\Exception\\NotNullConstraintViolationException(code: 0): An exception occurred while executing 'INSERT INTO task (repository_full_name, number, action, created_at, updated_at, verify_after) VALUES (?, ?, ?, ?, ?, ?)' with params [\"carsonbot-playground\\/symfony\", 22, 4, \"2021-05-31 14:10:11\", \"2021-05-31 14:10:11\", \"2021-06-01 10:10:11\"]:\n\nSQLSTATE[23502]: Not null violation: 7 ERROR:  null value in column \"id\" violates not-null constraint\nDETAIL:  Failing row contains (null, carsonbot-playground/symfony, 22, 4, 2021-05-31 14:10:11, 2021-05-31 14:10:11, 2021-06-01 10:10:11). at /app/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractPostgreSQLDriver.php:64)\n[previous exception] [object] (Doctrine\\DBAL\\Driver\\PDO\\Exception(code: 23502): SQLSTATE[23502]: Not null violation: 7 ERROR:  null value in column \"id\" violates

**Comment by Nyholm** [comment]
PR #181 fixed the issue. But Im not sure why..

**Comment by Nyholm** [comment]
Using doctrine/doctrine-bundle 2.3.2 is bad. The error exists in that version.

**Comment by Nyholm** [comment]
Using doctrine/doctrine-bundle 2.2.4 is okey.

**Comment by Nyholm** [comment]
Using doctrine/doctrine-bundle 2.3.0 is bad. This version is the first version that is broken.

**Comment by stof** [comment]
@Nyholm can you try configuring the db driver and server versions explicitly in the DBAL config instead of extracting them from the credentials URL ? DoctrineBundle 2.3.0 introduces a new metadata caching configuration that automatically cache the mapping during the cache-warmup (enabled only in non-debug mode, which might explain why you don't see the issue in dev). If you run the cache-warmup without the DB credentials in the env variables, the ORM might receive the wrong DB platform when generating the metadata, which would cache the wrong metadata (and so potentially break things)

**Comment by stof** [comment]
the .env file configures a sqlite URL, not a postgresql one. And the smart mapping defaults of the ORM (for `IdGenerator="auto" for instance) produce a different mapping on postgresql than on sqlite.

**Comment by Nyholm** [comment]
FYI: I can access the database. Reading is fine. 

Im using the following DSN: `postgres://xxx:yyy@database.internal:5432/my_db?sslmode=disable&charset=utf8&serverVersion=12`

**Comment by stof** [comment]
@Nyholm but is this DSN available in the env variable *at the time you run `cache-warmup` ? Or does it use the `sqlite` one from .env ?

**Comment by Nyholm** [comment]
@tucksaun, could you help me answer this question: 

> @Nyholm but is this DSN available in the env variable *at the time you run cache-warmup ? Or does it use the sqlite one from .env ?

**Comment by Nyholm** [comment]
Yes @stof. I think you are correct. I was able to reproduce this locally. 

https://github.com/symfony-tools/carsonbot/pull/189#issuecomment-853278391

So basically, warm the cache with `DATABASE_URL=sqlite:///%kernel.project_dir%/var/data.db`, then use `DATABASE_URL=postgres://foo:bar@127.0.0.1/carson` in production and you will have this issue. 

> @Nyholm but is this DSN available in the env variable *at the time you run cache-warmup ? Or does it use the sqlite one from .env ?

So from the testing, I would say: no. The environment variables are not available when we warm the cache.

**Comment by stof** [comment]
and so, the solution for that is either to put a postgresql URL in the `.env` for the fallback value, or maybe to set the driver explicitly in YAML (but maybe the fake URL from `.env` will still win, not sure).

**Comment by tucksaun** [comment]
@Nyholm I suppose you pinged me because this happens when using SymfonyCloud, right?
If I get the issue correctly from what @stof said, this looks similar to [the issue with the `server_version `](https://symfony.com/doc/current/cloud/troubleshooting.html#sqlstate-hy000-2002-connection-refused-during-the-build-step) (tl;dr: during the cache warmup, Doctrine needs to connect to the database to determine the platform version). 

If that is the case, this will indeed not be possible on SymfonyCloud because the database is not available when the cache is first warmed up (at build time). Technically, the database driver can not be inferred at build time (the service type can be changed from one deploy to the another so it's decoupled from the build) but you would have to confirm this with SymfonyCloud support (I no longer work for them).

But I guess it just makes more sense to have a default database URL in `.env` that matches the production driver.

**Comment by stof** [comment]
@tucksaun the ORM metadata cache depends on the DBAL platform (which depends on the driver and server version). It does not need to connect to the DB if the server version is provided in the config rather than being inferred.
Changing the database type without a build won't work, unless you don't do the cache warmup during the build.

**Comment by tucksaun** [comment]
@stof yes you are right, sorry I was too fast writing this 🙇 I was only meaning the reason was analogous.

@Nyholm it looks like you manage to tackle the issue right?

**Comment by Nyholm** [comment]
Yes, Im fine. 
But I fear many more doctrine users will have this issue.

**Comment by stof** [comment]
@tucksaun if you expect SymfonyCloud to allow switching from MySQL to Postgresql (or the opposite) without rebuilding the Symfony cache (which is what I understand from your message), then SymfonyCloud is incompatible with Doctrine ORM (or at least incompatible with an optimized ORM setup that warms up the metadata cache, which is what recent DoctrineBundle versions do by default)

**Comment by tucksaun** [comment]
@stof I'm not saying this is expected but this is possible to do so because credentials are injected at runtime. You can think about it as container warmup done during a Docker build for example. This allows SymfonyCloud (and PSH) to reuse one "image" from one deployment to the other, for example if you switch branches or upgrade database version.

@Nyholm Yeah you are probably right. Though I would think most users will tend to have a default DSN in `.env` using the same driver as production. You probably want to reach SymfonyCloud team so that they upgrade their documentation to mention it.

**Comment by Nyholm** [comment]
PR #192 fixed the issue

**Comment by stof** [comment]
@Nyholm this issue should be closed
